features/323-login-id/design.md

```diff
@@ -0,0 +1,253 @@
+    - Password authentication: Current none
+    - OAuth: (OAuth token, origin profile)
+    - Custom token: Current none
+Principal identity should be immutable, and principal metadata should be updated
+only through identity-provider specific APIs.
+
+
+To better model this design, the authn response should contains the principal instead.
+i.e.
+```json
+{
+    "access_token": "...",
+    "user": {
+        "id": "..."
+    },
+    "principal": {
+        "id": "...",
+        "realm": "teacher",
+        "email": "a@example.com",
+        "metadata": {}
+    }
+}
+```
+
+All auth APIs should be revised so that principal ID is used when appropiate.
+
+To update login ID, API consumer should instead perform a series of operations:
+- Create password authenticated principal with new login ID.
+- Ensure new principal is usable. (e.g. verified email/phone, approved by admin)
+- Delete the principal of old login ID.
+
+For passcodes (#293), a new principal associated with a specical identity
+provider should be used instead.
+
+The principal manipulation APIs would be specified in another specification.
+
+This address use case 1, 6.
+
+## Allow Multiple Login ID per Login ID Key
+In some website, user can associate multiple emails (potentially unlimited) with
+same user. These email may be used in auth features:
+- Login with multiple emails
+- Backup email
+- Welcome email
+
+A login ID key should allow having multiple associated login ID.
+When signing up, user should provide a list of login ID. e.g.
+```json
+{
+    "login_ids": [
+        { "username": "kiootic" },
+        { "email": "kiootic+1@oursky.com" },
+        { "email": "kiootic+2@oursky.com" }
+    ],
+    "password": "12345678"
+}
+```
+
+This addresses use case 6.
+
+
+## Standard Semantics Keys
+With reference to OpenID Connect standard claims, a list of standard keys
+would be specified:
+- `username`
+- `email`
+- `phone`
+- `name`
+- `picture`
+- `locale`
+- `zoneinfo`
+
+These keys would be used in principal metadata.
+
+Password Recovery & 2FA can use email/phone would be able to use
+information in principal metadata directly.
+
+Identity provider is responsible to populate principal metadata with appropiate
+information.
+
+
+## Configurable Login ID Keys
+The login ID keys are whitelisted by developer. In additional, developer can
+configure each login ID key:
+- The corresponding standard semantics key (normalization would be applied)
+- The minimum/maximum amount of login ID
+
+The password identity provider would return principal metadata according to
+the configuration.
+
+
+## User Metadata
+As specified in #256, auth gear stores 'user metadata'. However this name is
+misleading: this piece of data is technically NOT metadata.
+
+The current 'user metadata' would be renamed as 'user custom data'. Auth gear
+does not concern the content of user custom data.
+
+A new 'user metadata' would be added. This metadata is a map of strings,
```

Comment:
Does it meant the "metadata" in the user object would only allow the following two keys? If yes please be specific about it.

BTW, in that case, I don't think we should define two new terminologies `metadata` and `customdata`, it is too confusing. If it is only two keys I think we can just put it in user object?

The reason is to avoid introducing too many concepts which makes understanding of Skygear difficult for new comers / future developers...